rpc/src/server/service.rs

```diff
@@ -663,229 +687,135 @@ mod fns {
+
+                let level_endorsing_rights = complete_endorsing_rights_for_level(context_data, parameters, constants, level, level, estimated_time)?;
+                endorsing_rights = merge_slices!(&endorsing_rights, &level_endorsing_rights);
+            }
         } else {
-            requested_level..requested_level+1
+            // use level prepared during parameter parsing to compute estimated time
+            let estimated_time: Option<String> = parameters.get_estimated_time(constants, None);
+
+            endorsing_rights = complete_endorsing_rights_for_level(context_data, parameters, constants, *parameters.requested_level(), *parameters.display_level(), estimated_time)?;
         };
 
-        for level in level_iterator {
-            //check if estimated time is computed and convert from raw epoch time to rfc3339 format
-            let timestamp_level = if is_cycle {
-                level-1
-            } else {
-                timestamp_level
-            };
+        Ok(Some(RpcResponseData::EndorsingRights(endorsing_rights)))
+    }
+    
+    /// Use prepared data to generate endosring rights
+    /// 
+    /// # Arguments
+    /// 
+    /// * `context_data` - Data from context list used in baking and endorsing rights generation filled in [RightsContextData](RightsContextData::prepare_context_data_for_rights).
+    /// * `parameters` - Parameters created by [RightsParams](RightsParams::parse_rights_parameters).
+    /// * `constants` - Context constants used in baking and endorsing rights [`get_and_parse_rights_constants`].
+    /// * `level` - Level to feed Tezos PRNG.
+    /// * `display_level` - Level to be displayed in output.
+    /// * `estimated_time` - Estimated time of endorsement, is set to None if in past relative to block_id.
+    #[inline]
+    fn complete_endorsing_rights_for_level(context_data: &RightsContextData, parameters: &RightsParams, constants: &RightsConstants, level: i64, display_level: i64, estimated_time: Option<String>) -> Result<Vec::<EndorsingRight>, failure::Error> {
+        let mut endorsing_rights = Vec::<EndorsingRight>::new();
 
-            let estimated_time: Option<String> = if block_level <= timestamp_level {
-                let est_timestamp = ((timestamp_level - block_level).abs() as i64 * time_between_blocks[0]) + block_timestamp;
-                Some(ts_to_rfc3339(est_timestamp))
-            } else {
-                None
-            };
+        // endorsers_slots is needed to group all slots by delegate
+        let endorsers_slots = get_endorsers_slots(constants, context_data, level)?;
 
-            // endorsers_slots is needed to group all slots by delegate
-            let endorsers_slots = get_endorsers_slots(&constants, &cycle_data, level)?;
+        // order descending by delegate public key hash address hex byte string
+        for delegate in endorsers_slots.keys().sorted().rev() {
+            let delegate_data = endorsers_slots.get(delegate).unwrap();
```

Comment:
HashMap not containing an key, is expected scenario. Replace `unwrap` with returning `Err(...)`. (There is convenience method for that too: [Option::ok_or](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or)).